File: week03/ohy/배달.java
import java.util.*;

class Solution {

    static class Node {
        private final int idx;
        private final int cost;

        Node(int idx, int cost) {
            this.idx = idx;
            this.cost = cost;
        }
    }

    public int solution(int N, int[][] road, int K) {
        int answer = 0;

        ArrayList<ArrayList<Node>> graph = new ArrayList<>();
        
        for (int i = 0; i <= N; i++) {
            graph.add(new ArrayList<>());
        }
        
        for (int i = 0; i < road.length; i++) {
            int s = road[i][0];
            int e = road[i][1];
            int c = road[i][2];
            
            graph.get(s).add(new Node(e, c));
            graph.get(e).add(new Node(s, c));
        }
        
        int[] dist = new int[N + 1];
        Arrays.fill(dist, Integer.MAX_VALUE);
        dist[1] = 0;

        PriorityQueue<Node> pq = new PriorityQueue<>((o1, o2) -> Integer.compare(o1.cost, o2.cost));
        pq.offer(new Node(1, 0));
        
        while (!pq.isEmpty()) {
            Node current = pq.poll();
            
            if (current.cost > dist[current.idx]) continue;
            
            for (Node next : graph.get(current.idx)) {
                int cost = current.cost + next.cost;
                
                if (cost < dist[next.idx]) {
                    dist[next.idx] = cost;
                    pq.offer(new Node(next.idx, cost));
                }
            }
        }

        for (int i = 1; i <= N; i++) {
            if (dist[i] <= K) answer++;
        }
        return answer;
    }
}

File: week03/kjg/배달.js
function solution(N, road, K) {
    
    const dist = Array(N+1).fill(Infinity) // 출발지로부터 거리
    const visited = Array(N+1).fill(false) //방문
    const graph = Array.from({length: N+1},()=> Array(N+1).fill(Infinity)) //각 정점간 거리


    //graph 채우기
    for(const [a,b,time] of road){ // graph세팅 및 두 정점간 복수개 간선 최소간선값으로 단일화
        if(time < graph[a][b]){
            graph[a][b] = time
            graph[b][a] = time
        }
    }

    dist[1] = 0 // 시작점 1 거리 0 / 확정값 

    for(let i = 0; i<N; i++){
        let minNode = 0
        let minTime = Infinity

        for(let j = 1;j < N+1; j++){
            if(!visited[j] && dist[j] < minTime){ // 방문하지않은 min확정값인가
                minTime = dist[j];
                minNode = j
            }
        } 
        // 방문하지않은 확정값중 최소확정값 확정됨

        if(minNode === 0) break;

        // 방문하지않은 최소확정값 노드 방문처리
        visited[minNode] = true

        for(let next = 1; next < N+1; next++){
            if(graph[minNode][next] !== Infinity){ // 방문처리된것과 이어진 정점
                const newDist = dist[minNode] + graph[minNode][next]
                if(newDist < dist[next]){// 신규 간선값이 더 작으면 업데이트
                    dist[next] = newDist
                }
            }
        }


    }
    return dist.filter(d => d <= K).length;
}
